### ISSUE TITLE
trace_set_err call with errno when errno might be zero

### ISSUE THREAD

**Issue by razeh** [LibtraceTeam/libtrace#56]
In format_pcapfile.c:355 we have:
```
        err=wandio_read(libtrace->io,
                        packet->buffer,
                        sizeof(libtrace_pcapfile_pkt_hdr_t));
        if (err<0) {
                trace_set_err(libtrace,errno,"reading packet");
                return -1;
        }
        if (err==0) {
                /* EOF */
                return 0;
        }
```
The problem is that wandio_read can return an error without errno being set (or at least that would explain what I'm seeing). Inside of trace_set_err there is an assert that the errcode is non-zero. This turns an error that I'd like to recover from into an assertion failure, which I cannot recover from.

**Comment by rsanger** [comment]
Thanks for the bug report

Is it possible to get some more information on what caused this issue?
What file type and compression were you using? Is this something that you can easily replicate?
If possible a backtrace from gdb would be helpful, or the file causing issues.

This looks like something that needs to be fixed in wandio, to ensure it always returns an errno.

In terms of preventing this issue again in libtrace, I'm not sure what the best long term solution to this issue is.
Maybe replacing the assertion with an error message, asking this to be reported as a bug?
Thoughts on this @salcock @alistairking ?

**Comment by salcock** [comment]
Agree with pretty much everything you've said.

Libtrace really shouldn't be assert failing in these situations. We tend to be guilty of putting asserts in during development to make sure we are alerted when something doesn't quite go as planned and then never removing / replacing them once the code is finished.

In the case of trace_set_err, this is probably a case where the set of built in libtrace error codes needs to be expanded to cover more situations. A TRACE_ERR_WANDIO_FAILED (if it existed) would be more appropriate to pass in here than errno.

Ideally, wandio would also provide some sort of error output that would be helpful to any user trying to interpret what went wrong but that is a slightly separate issue :)

**Comment by razeh** [comment]
Here is a call stack:
#0  0x00002aaaae26d1d7 in raise () from /lib64/libc.so.6
#1  0x00002aaaae26e8c8 in abort () from /lib64/libc.so.6
#2  0x00002aaaae266146 in __assert_fail_base () from /lib64/libc.so.6
#3  0x00002aaaae2661f2 in __assert_fail () from /lib64/libc.so.6
#4  0x000000000110cc19 in trace_set_err ()
#5  0x000000000110d9d0 in pcapfile_read_packet ()
#6  0x0000000001100ddd in trace_read_packet ()

I will work on creating a TRACE_ERR_WANDIO_FAILED and see if that solves the problem for me. 
I suspect the problem is a broken or partial capture; unfortunately I can't provide the capture. I will try to narrow down what's going wrong.

**Comment by razeh** [comment]
I made a pull request with my fix: https://github.com/LibtraceTeam/libtrace/pull/57

**Comment by salcock** [comment]
Thanks for that. We'll try to review it shortly.

**Comment by salcock** [comment]
Original issue should be resolved. Still probably worth reviewing the rest of our error handling and removing unnecessary asserts at some point, but that's a job for another day.
